src/com/t_oster/liblasercut/drivers/K40NanoDriver.java

```diff
@@ -0,0 +1,1484 @@
+   * @throws IllegalJobException Throw this exception, when the job is not
+   * suitable for the current machine
+   * @throws Exception
+   */
+  @Override
+  public void sendJob(LaserJob job, ProgressListener pl, List<String> warnings) throws IllegalJobException, Exception
+  {
+    //let's check the job for some errors
+    checkJob(job);
+    K40Device device = new K40Device();
+    device.setBoard(board);
+
+    device.open();
+
+    //Well, first, let's iterate over the different parts of this job.
+    for (JobPart p : job.getParts())
+    {
+      //now we have to check, of which kind this part is. We only accept VectorParts and add a warning for other parts.
+      if (!(p instanceof VectorPart))
+      {
+        warnings.add("Non-vector parts are ignored by this driver.");
+      }
+      else
+      {
+        //so, we know it's a VectorPart. We cast it, so we get the real interface
+        VectorPart vp = (VectorPart) p;
+        //A VectorPart consists of a command List. So let's iterate over this list
+        for (VectorCommand cmd : vp.getCommandList())
+        {
+          //There are three types of commands: MOVETO, LINETO and SETPROPERTY
+          switch (cmd.getType())
+          {
+            case LINETO:
+            {
+              /**
+               * Move the laserhead (laser on) from the current position to the
+               * x/y position of this command. All coordinates are in dots
+               * respecting to the job resolution
+               */
+              int x = (int) (Util.mm2inch(Util.px2mm(cmd.getX(), p.getDPI())) * 1000.0);
+              int y = (int) (Util.mm2inch(Util.px2mm(cmd.getY(), p.getDPI())) * 1000.0);
+              //Native units are mils.
+              device.cut_absolute(x, y);
+              device.execute();
+              break;
+            }
+            case MOVETO:
+            {
+              /**
+               * Move the laserhead (laser off) from the current position to the
+               * x/y position of this command. All coordinates are in mm
+               */
+              int x = (int) (Util.mm2inch(Util.px2mm(cmd.getX(), p.getDPI())) * 1000.0);
+              int y = (int) (Util.mm2inch(Util.px2mm(cmd.getY(), p.getDPI())) * 1000.0);
+              //Native units are mils.
+              device.move_absolute(x, y);
+              device.execute();
+              break;
+            }
+            case SETPROPERTY:
+            {
+              /**
+               * Change properties of current laser-actions (e.g. speed,
+               * frequency, power... whatever your driver supports)
+               */
+              LaserProperty prop = cmd.getProperty();
+              for (String key : prop.getPropertyKeys())
+              {
+                String value = prop.getProperty(key).toString();
+                if ("speed".equals(key))
+                {
+                  device.setSpeed(Double.valueOf(value));
```

Comment:
if I understand correctly, you're currently using the percentage value (speed 0...100) as mm/s, so there's a maximum of 100. That's okay -- choose as you like -- but should be documented in the code

Note that if you later change the meaning of "100", it will break existing saved laser settings, so you should rather decide now instead of in two years when everyone already has a list of saved material settings :-)

Reply:
Yeah, I was working to plug in pieces where they seemed to go and finish up the driver elements. I mostly skipped that part and used the default even though there's clearly other elements that need to go in there and bits that shouldn't be used.